### PR TITLE
IOBUF support

### PR DESCRIPTION
**Pull request by mkurc-ant**
This PR adds support for IOBUF to SymbiFlow.

- [x] Minitest.
- [x] Blackbox / simulation model.
- [x] Architecture definition.
- [x] Techmap.
- [x] fasm2bels.
- [ ] Automatic inference of 3-state IOBUFs

With this PR the flow of SymbiFlow changes a bit. Upon sysnthesis the design is written to a JSON file. Then, the file is processed by the `split_inouts.py` utility which splits inout ports into separate inputs and outputs. After processing, the design is written to another JSON file which is then converted to EBLIF using Yosys.

This processing should be moved to a Yosys plugin in the future.

Each top-level inout port of the design is split into one input and one output port. Port names are given in the following way: `inout A` -> `input A_$inp` and `output A_$out`.

### CODE REVIEW

**Comment by mkurc-ant** [comment]
I'm still having issues with making Yosys infer IOBUFs from assign statements though. For such a design:
```
inout wire pin

...

wire pin_i;
wire pin_o;
wire pin_t;

assign pin_i = pin;
assign pin = (pin_t == 1'b1) ? pin_o : 1'bz;
```

In some cases I end up with an `OBUFT`-like cell for the 3-state output and the input taken directly from the pin (not through any buffer). In other cases I have a properly connected `IOBUF` cell but after optimization its output gets disconnected (a bug?).

**Comment by litghost** [comment]
> I'm still having issues with making Yosys infer IOBUFs from assign statements though. For such a design:

I know I suggested the iobufmap to begin with, but if we simply require the IOBUF explicitly, would your script do the right thing?

File: common/cmake/devices.cmake [hunk 1]
```diff
@@ -12,7 +12,8 @@ function(DEFINE_ARCH)
   # ~~~
   # DEFINE_ARCH(
   #    ARCH <arch>
-  #    YOSYS_SCRIPT <yosys_script>
+  #    YOSYS_SYNTH_SCRIPT <yosys_script>
+  #    YOSYS_CONV_SCRIPT <yosys_script>
```
Comment (litghost) [inline common/cmake/devices.cmake]: Need to update documentation on these arguments

File: common/cmake/devices.cmake [hunk 2]
```diff
@@ -980,31 +985,62 @@ function(ADD_FPGA_TARGET)
...
       VERBATIM
     )
+
+    set(SPLIT_INOUTS ${symbiflow-arch-defs_SOURCE_DIR}/utils/split_inouts.py)
+
+    add_custom_command(
+      OUTPUT ${OUT_JSON}
+      DEPENDS ${OUT_JSON_SYNTH} ${QUIET_CMD} ${QUIET_CMD_TARGET} ${SPLIT_INOUTS} ${PYTHON3} ${PYTHON3_TARGET} simplejson
+      COMMAND
+        ${CMAKE_COMMAND} -E env PYTHONPATH=${PRJXRAY_DIR}:${symbiflow-arch-defs_SOURCE_DIR}/utils ${PYTHON3} ${SPLIT_INOUTS} -i ${OUT_JSON_SYNTH} -o ${OUT_JSON}
```
Comment (litghost) [inline common/cmake/devices.cmake]: What is PRJXRAY_DIR required here?   It seems weird to have a 7-series specific thing in the (in theory) generic CMake function?

File: utils/split_inouts.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+            return n0 + 1
+
+    # No gap was found, return max + 1.
+    return sorted_nets[-1] + 1
+
+
+# =============================================================================
+
+
+def find_and_split_inout_ports(design, module_name):
```
Comment (litghost) [inline utils/split_inouts.py]: Docstring

File: utils/split_inouts.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        attrs = module["attributes"]
+        if "top" in attrs and attrs["top"] == 1:
+            return name
+
+    return None
+
+
+def get_nets(bits):
+    """
+    Returns a set of numbers corresponding to net indices.
```
Comment (litghost) [inline utils/split_inouts.py]: This function is very weird.  Can you provide some doctest examples?

File: utils/split_inouts.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    """
+    Looks for the top-level module in the design. Returns its name.
+    """
+
+    for name, module in design["modules"].items():
+        attrs = module["attributes"]
+        if "top" in attrs and attrs["top"] == 1:
+            return name
+
+    return None
```
Comment (litghost) [inline utils/split_inouts.py]: Should throw an exception rather than return None?

File: utils/split_inouts.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+def get_nets(bits):
+    """
+    Returns a set of numbers corresponding to net indices.
+    """
+    return set([n for n in bits if isinstance(n, int)])
+
+
+def get_free_net(nets):
+    """
+    Given a set of used net indices, returns a new, free index.
```
Comment (litghost) [inline utils/split_inouts.py]: This explaination is insufficient to understand what you are doing.  Needs more.

File: utils/split_inouts.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+# =============================================================================
+
+
+def find_and_split_inout_ports(design, module_name):
+
+    # Get the module
+    module = design["modules"][module_name]
+
+    # Find indices of all used nets
```
Comment (litghost) [inline utils/split_inouts.py]: I think you need to document the income JSON format here in the comment (or at least link to other documentation)

File: utils/split_inouts.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        netnames[name] = {
+            "hide_name": 0,
+            "bits": port["bits"],
+            "attributes": {}
+        }
+
+    return port_map, net_map
+
+
+def remap_connections(design, module_name, net_map):
```
Comment (litghost) [inline utils/split_inouts.py]: Docstring

File: common/cmake/devices.cmake (hunk 2 above)
Comment (litghost) [reply common/cmake/devices.cmake]: Upon review of split_inouts, it doesn't seem to use PRJXRAY at all? So you should just be able to drop the path?

File: common/cmake/devices.cmake (hunk 1 above)
Comment (mkurc-ant) [reply common/cmake/devices.cmake]: Done.

File: common/cmake/devices.cmake (hunk 2 above)
Comment (mkurc-ant) [reply common/cmake/devices.cmake]: Right, It ended up here as a mistake. Removed.

File: utils/split_inouts.py (hunk 3 above)
Comment (mkurc-ant) [reply utils/split_inouts.py]: Done

File: utils/split_inouts.py (hunk 4 above)
Comment (mkurc-ant) [reply utils/split_inouts.py]: Added docstring and doctest.

File: utils/split_inouts.py (hunk 5 above)
Comment (mkurc-ant) [reply utils/split_inouts.py]: Added throwing an exception.

File: utils/split_inouts.py (hunk 6 above)
Comment (mkurc-ant) [reply utils/split_inouts.py]: Added docstring and doctest.

File: utils/split_inouts.py (hunk 8 above)
Comment (mkurc-ant) [reply utils/split_inouts.py]: Added.

File: utils/split_inouts.py (hunk 7 above)
Comment (mkurc-ant) [reply utils/split_inouts.py]: I linked up Yosys documentation to the docstring of the file. Couldn't find any "official" documentation of the JSON design format other than Yosys' manual for the `write_json` command.

**Pull request merged**